[DO NOT REVIEW] [WIP] Update codegen to support ScalarList

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44519 [WIP] Add foreach APIs with ScalarList
* **#44518 Update codegen to support ScalarList**

Foreach APIs require ability to pass a list of scalars as an input. This PR enables passing scalar lists as a parameter. 

Differential Revision: [D23699510](https://our.internmc.facebook.com/intern/diff/D23699510)

Commit:
initial

[ghstack-poisoned]

Commit:
Update on "[WIP] Update codegen to support ScalarList"




[ghstack-poisoned]

ezyang commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -48,6 +48,7 @@ namespace at {
 
 class Tensor;
 using TensorList = ArrayRef<Tensor>;
+using ScalarList = ArrayRef<Scalar>;
```
This is a minor point (and implies some divergence from TensorList), but I'd prefer it if we stopped defining so many aliases for these types. ScalarList is doubly bad because it implies an owning list (which an ArrayRef is not). You don't need the typedef unless you actually translate `Scalar[]` into `ScalarList`, and preventing this translation is as simple as removing a case from the codegen...

ezyang commented on tools/codegen/api/cpp.py:
```diff
@@ -99,6 +99,8 @@ def argumenttype_type(t: Type, *, mutable: bool) -> str:
         elif str(t.elem) == 'Tensor':
             return "TensorList"
+        elif str(t.elem) == 'Scalar':
+            return "ScalarList"
```
...that case is right here.

ezyang replied:
I do admit it would be easier to do this if the autograd codegen was also consistent on this matter, but I suspect that a search replace on ScalarList to ArrayRef<ScalarList> ought to be enough.

ezyang replied:
To spell out this suggestion in more detail. In this code, you have translated the type `Scalar[]` (that's a ListType of Scalar BaseType) into `ScalarList`, by cargo-culting the code above. Above however, there is a comment:

```
# TODO: remove these special cases, ArrayRef fallthrough works fine
```

What this comment is saying is that the default behavior below:

```
return f"ArrayRef<{elem}>"
```

would work perfectly well as a translation of this type into a C++ type: you would get `ArrayRef<Scalar>` instead of `ScalarList`, which are the same thing because elsewhere in this PR, you had to define ScalarList to be `ArrayRef<Scalar>` to make this translation work out.

We should prefer directly using `ArrayRef<Scalar>` in our C++ API, for two reasons: (1) it clearly specifies that the array in question is a *reference* which means it shouldn't be treated as a value type (TensorList also has this same problem, and it is why we renamed IntList to IntArrayRef), (2) it reduces the number of special casing and extra conventions we need. We have this implicit convention that a ton of people have cargo culted that you have to define a type alias for every list type you need, and that increases the amount of boilerplate you have to write in this case.

Once you do this, you will have to fix downstream usages of ScalarList to now match for ArrayRef<Scalar>. But this should be easy to do, as you have already identified all such occurrences.

Commit:
Update on "[WIP] Update codegen to support ScalarList"


Foreach APIs require ability to pass a list of scalars as an input. This PR enables ScalarList object for that. 

[ghstack-poisoned]

Commit:
Update on "[WIP] Update codegen to support ScalarList"


Foreach APIs require ability to pass a list of scalars as an input. This PR enables ScalarList object for that. 

[ghstack-poisoned]
